### ISSUE TITLE
What's the state of this?

### ISSUE THREAD

**Issue by ciriousjoker** [qinghua921/ortools#1]
Hey!
We're a really small company and we're trying to solve a CSP that or-tools can apparently help us solve. A week ago I started looking for solutions for how to use or-tools inside Typescript. Right now we don't need something like this yet(!), but that point in time will inevitably come pretty soon, so I guess I'm really lucky that you started creating bindings a few days ago :D

### I'm currently wondering a few things
1. Who's working on this? Just you? Is there any company backing you?
2. I'm a bit confused. Who are you? You've had one contribution on your Github timeline in 2015, then another day with 3 contributions in 2023 and a few days ago you suddenly started heavily pushing code.
3. What's the purpose for this? I'm talking about the purpose specifically for you, obviously having or-tools in Typescript is amazing by itself, but I'm wondering if this thing will only exist to solve one issue you have and then be dropped.
4. This seems to be actively developed, at least right now. Are you expecting this thing to die after it served its purpose?
5. Is there a plan for how this could be maintained in the long run, after Google made its changes? I'd assume that all you'd have to do is mirror their function calls in C++ and cmake creates the bindings automatically? That seems like relatively easy to maintain, but I have no clue
6. Any plans to merge this upstream into or-tools? People have wanted Javascript support for literally 10 years at this point: https://github.com/google/or-tools/issues/94
7. As for us, we're specifically interested in the CP-SAT solver (at least to my current knowledge). Do you think the bindings for this will be ready within the next month or so?
8. Do you accept PRs? I currently have no absolutely clue how or-tools or node's native binding system work, but if you're open to it I might look into it :)

### TLDR
I'm trying to figure out if we can rely on this package to solve our problem and be available in the future. It kind of seems too good to be true right now.

**Comment by qinghua921** [comment]
**Firstly, to answer your questions:**

1. **Who is working on this? Just you? Is there any company backing you?**
   - It's just me working on this code; there is no company backing it.

2. **I’m a bit confused. Who are you? You've had one contribution on your GitHub timeline in 2015, then another day with 3 contributions in 2023 and a few days ago you suddenly started heavily pushing code.**
   - This project is actually my personal contribution to open source, and it might also be my last contribution. I'm no longer working as a programmer; this project is just a matter of personal interest for me.

3. **What's the purpose for this? I'm talking about the purpose specifically for you, obviously having or-tools in TypeScript is amazing by itself, but I'm wondering if this thing will only exist to solve one issue you have and then be dropped.**
   - This project is not just to solve my own problem and will not be abandoned halfway. However, updates might not be very frequent; I will try my best to keep it updated.

4. **This seems to be actively developed, at least right now. Are you expecting this thing to die after it served its purpose?**
   - Ideally, I hope this project can be maintained in the long term. The current mission is to complete the bindings for the 2019 version of or-tools. Future versions will depend on whether this project proves to be valuable and if there is significant usage.

5. **Is there a plan for how this could be maintained in the long run, after Google made its changes? I'd assume that all you'd have to do is mirror their function calls in C++ and cmake creates the bindings automatically? That seems like relatively easy to maintain, but I have no clue.**
   - There is no long-term plan at the moment. Most of the time, it’s about creating simple bindings, but some parts of the code do require more thought. In fact, the main challenge of this project is organizing the code and addressing the differences between JavaScript and C++.

6. **Any plans to merge this upstream into or-tools? People have wanted JavaScript support for literally 10 years at this point: Is there any plans for JavaScript bindings for or-tools? [google/or-tools#94](https://github.com/google/or-tools/issues/94)**
   - There are currently no plans to merge this into the main or-tools repository. If you specifically need the CP-SAT solver, I can prepare it in advance and will try to help as much as possible.

7. **As for us, we're specifically interested in the CP-SAT solver (at least to my current knowledge). Do you think the bindings for this will be ready within the next month or so?**
   - I accept PRs. In fact, most of the time, the code is quite simple or even a bit mundane.

**Now, to introduce the project, its possible future, and issues:**

1. **This project has only been written on Windows so far; it hasn't been tested on Linux.**
2. **The organization of the C++ code may change; I am thinking about how to better organize the code. The TypeScript interface might have some small changes, but they will not be major.**
3. **Currently, it’s based on the 2019 version of or-tools, not Google’s latest version. It’s challenging for a single person to keep up with a team’s pace.**
4. **The main goal at this stage is to complete the C++ demo. Unit tests have not kept up, so there might be bugs. Be cautious when using it, especially with potential hard-to-detect memory leaks. If there are issues, please feel free to open an issue, and I'll resolve it.**

**Comment by DimuDesigns** [comment]
@qinghua921 Appreciate your efforts. I have a question. Have you considered porting OR Tools to Web Assembly instead of using a wrapper? Maybe that would be easier to maintain over time.

**Comment by qinghua921** [comment]
Yes, in fact, I have also noticed that a significant amount of the code is quite similar. I am currently considering using SWIG, WebAssembly, or Clang to automatically generate the code, and I am comparing the pros and cons of these three approaches.

**Comment by ciriousjoker** [comment]
@qinghua921 After some digging, I've realized that the CP-SAT solver (specifically cp_model.CpSolver from the python version) is highly likely to solve our problem. Is there any way you could prioritize the development of this?

**Comment by qinghua921** [comment]
@ciriousjoker ok
